cmd/query/app/server.go

```diff
@@ -131,7 +123,67 @@ func createGRPCServer(querySvc *querysvc.QueryService, metricsQuerySvc querysvc.
 	healthServer.SetServingStatus("jaeger.api_v3.QueryService", grpc_health_v1.HealthCheckResponse_SERVING)
 
 	grpc_health_v1.RegisterHealthServer(server, healthServer)
-	return server, nil
+}
+
+func createGRPCServer(
+	ctx context.Context,
+	legacy bool,
+	querySvc *querysvc.QueryService,
+	metricsQuerySvc querysvc.MetricsQueryService,
+	options *QueryOptions,
+	tm *tenancy.Manager,
+	telset telemetery.Setting,
+) (*grpc.Server, error) {
+	var grpcServer *grpc.Server
+	if legacy {
+		var grpcOpts []grpc.ServerOption
+		if options.GRPC.TLSSetting != nil {
+			tlsCfg, err := options.GRPC.TLSSetting.LoadTLSConfig(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			creds := credentials.NewTLS(tlsCfg)
+
+			grpcOpts = append(grpcOpts, grpc.Creds(creds))
+		}
+		if tm.Enabled {
+			grpcOpts = append(grpcOpts,
+				//nolint
+				grpc.StreamInterceptor(tenancy.NewGuardingStreamInterceptor(tm)),
+				grpc.UnaryInterceptor(tenancy.NewGuardingUnaryInterceptor(tm)),
+			)
+		}
+
+		grpcServer = grpc.NewServer(grpcOpts...)
+	} else {
+		var grpcOpts []configgrpc.ToServerOption
+		if tm.Enabled {
+			grpcOpts = append(grpcOpts,
+				//nolint
+				configgrpc.WithGrpcServerOption(grpc.StreamInterceptor(tenancy.NewGuardingStreamInterceptor(tm))),
+				configgrpc.WithGrpcServerOption(grpc.UnaryInterceptor(tenancy.NewGuardingUnaryInterceptor(tm))),
```

Comment:
why do we need the wrappers like `grpc.UnaryInterceptor()`? Isn't that what the New functions already return?

Reply:
Are you talking about the `ToServer` function setting this automatically? I actually wanted to ask you about that. This is the implementation in OTEL. The interceptors look slightly different so I wasn't sure if we still needed the ones defined here.

```go
	var uInterceptors []grpc.UnaryServerInterceptor
	var sInterceptors []grpc.StreamServerInterceptor

	if gss.Auth != nil {
		authenticator, err := gss.Auth.GetServerAuthenticator(context.Background(), host.GetExtensions())
		if err != nil {
			return nil, err
		}

		uInterceptors = append(uInterceptors, func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp any, err error) {
			return authUnaryServerInterceptor(ctx, req, info, handler, authenticator)
		})
		sInterceptors = append(sInterceptors, func(srv any, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
			return authStreamServerInterceptor(srv, ss, info, handler, authenticator)
		})
	}
```

Reply:
ah, nevermind, it's the 2-level options wrapping.